Workspaces api docgen source

ngrayluna opened the pull request:
## Summary of changes
* Adds descriptions to Reports API
* Hides some class methods that should not be exposed e.g. `to_model` and `from_model` using single underscores.  
   * This ensures that `lazydocs` can generate markdown files properly.
   *  (You can check they are exposed today by doing `help wandb_workspaces.reports.v2.ClassName` or `wandb_workspaces.reports.v2.ClassName?` in a notebook for classes that have either defined:
```   
def to_model()
```
or 
```
def from_model() 
```

* Updated reports and workspaces tests files to reflect `to_model `-> `_to_model` changes
* Removed `@classmethod` decorator on `Workspaces.save()` and `Workspaces.save_as_new`. Re: for the latter, the method is an instance method, not a class method.
* For those descriptions I was not sure of, I used CoPilot.
* Added "INSERT" for some (what I assume) are Classes that we don't want to show. Re: API generator sees them. In the future, do a) confirm and b) add an underscore so lazydocs doesn't pick them up. 

#### @anandwandb and/or someone to verify:
1. Should these be exposed? 
- [ ] `Spotify`
- [ ] `SoundCloud`
- [ ] `Twitter`

2. Do these `Weave*` Classes should be renamed to QueryPanels? i.e.
- [ ]  'WeaveBlockArtifact',
- [ ]  'WeaveBlockArtifactVersionedFile',
- [ ]  'WeaveBlockSummaryTable',
- [ ]  'WeavePanel',
- [ ]  'WeavePanelArtifact',
- [ ]  'WeavePanelArtifactVersionedFile',
- [ ]  'WeavePanelSummaryTable',

3. Finish descriptions for 
   a. `running_*` in `ScatterPlot`
   b. `gradient`  in `ParallelCoordinatesPlot`
   c.` chart_*` in `from_table`
   d. Overview description for for `Block` class


## Testing
* Passes circleCI tests. 
```
poetry run pytest --cov
```
* Unsure how to resolve, troubleshoot pre commit errors (returned error is vague). I tried:
```bash
poetry run pre-commit run --all-files
```

## Doc link preview: 
Preview the docs here https://reports-api.docodile.pages.dev/ref/python/wandb_workspaces/reports/ and here https://reports-api.docodile.pages.dev/ref/python/wandb_workspaces/workspaces

ngrayluna committed:
Added arguments to Workspace class

ngrayluna committed:
Hid from_model and to_model, reformatted attribute descriptions.

ngrayluna commented:
Known missing Class descriptions: 

'BarPlot',
 'BlockQuote',
 'CalloutBlock',
 'CheckedList',
 'CheckedListItem',
 'CodeBlock',
 'CodeComparer',
 'Config',
 'CustomChart',
 'Gallery',
 'GradientPoint',
 'H1',
 'H2',
 'H3',
 'HorizontalRule',
 'Image',
 'InlineCode',
 'InlineLatex',
 'LatexBlock',
 'Layout',
 'LinePlot',
 'Link',
 'MarkdownBlock',
 'MarkdownPanel',
 'MediaBrowser',
 'Metric',
 'OrderedList',
 'OrderedListItem',
 'P',
 'PanelGrid',
 'ParallelCoordinatesPlot',
 'ParallelCoordinatesPlotColumn',
 'ParameterImportancePlot',
 'Report',
 'RunComparer',
 'Runset',
 'RunsetGroup',
 'RunsetGroupKey',
 'ScalarChart',
 'ScatterPlot',
 'SummaryMetric',
 'TableOfContents',
 'UnorderedList',
 'UnorderedListItem',
 'Video',
 'WeaveBlockArtifact',
 'WeaveBlockArtifactVersionedFile',
 'WeaveBlockSummaryTable',
 'WeavePanel',
 'WeavePanelArtifact',
 'WeavePanelArtifactVersionedFile',
 'WeavePanelSummaryTable',

ngrayluna committed:
WIP

ngrayluna committed:
Added more descriptions

ngrayluna committed:
Quick nits

ngrayluna committed:
Added dec to init + more formatting

ngrayluna committed:
v0

ngrayluna committed:
Filled i more blanks

ngrayluna committed:
Added more descriptions

ngrayluna committed:
ran precommits

ngrayluna committed:
Testing ci check

ngrayluna committed:
Fixed from model

ngrayluna committed:
Fixed dunder to model in layput

ngrayluna committed:
Updated test_workspaces

ngrayluna committed:
Actually saved changes

ngrayluna committed:
workspaces test, to model in FilterExpr

ngrayluna committed:
Hacky fix?

ngrayluna committed:
Added self to self_model

ngrayluna committed:
Undo

ngrayluna committed:
Undo

ngrayluna committed:
Testing removing classmethod decorator

ngrayluna committed:
Pre commit checks

ngrayluna committed:
Filled in Gallery*

ngrayluna committed:
Added more descriptions to save and save as new

ngrayluna commented:
Preview the docs here: https://reports-api.docodile.pages.dev/ref/python/workspaces/wandb_workspaces.reports.v2.interface

andrewtruong commented on wandb_workspaces/reports/v2/__init__.py:
````diff
@@ -1,3 +1,12 @@
+"""Python library for programmatically working with Weights & Biases Reports API.
+
+```python
+# How to import
+import wandb_workspaces.reports.v2
````
This should be example usage like

```
import wandb_workspaces.reports.v2 as wr

wr.Report(...).save()
```
etc

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
````diff
@@ -1,4 +1,11 @@
-"""Public interfaces for the Report API."""
+"""Python library for programmatically working with Weights & Biases Reports API.
+
+```python
+# How to import
+import wandb_workspaces.reports.v2
+```
+
+"""
````
Same as comment above

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -82,43 +89,85 @@ def _spec(self):
 
 @dataclass(config=dataclass_config, repr=False)
 class Layout(Base):
+    """The layout of a block in a report. Adjusts the size and position of the block.
+
+    Attributes:
+        x: The x position of the block.
+        y: The y position of the block.
+        w: The width of the block.
+        h: The height of the block.
+    """
+
```
this is panel, not block layout

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -82,43 +89,85 @@ def _spec(self):
 
 @dataclass(config=dataclass_config, frozen=True)
 class Config:
+    """A configuration for a metric.
```
More clearly: This is something that a user logged as

```py
run.config[name] = ...
```

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -156,8 +211,10 @@ class TextWithInlineComments(Base):
 
 @dataclass(config=dataclass_config, repr=False)
 class Heading(Block):
+    """A heading block. INSERT"""
```
Insert what?

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -174,13 +231,19 @@ def from_model(cls, model: internal.Heading):
 
 @dataclass(config=dataclass_config, repr=False)
 class H1(Heading):
+    """Creates an H1 HTML tag with the text specified.
```
Probably more clear to say it's an H1 Heading (like Markdown)

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -191,13 +254,20 @@ def to_model(self):
 
 @dataclass(config=dataclass_config, repr=False)
 class H2(Heading):
+    """Creates an H2 HTML tag with the text specified.
```
Same as above

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -208,13 +278,20 @@ def to_model(self):
 
 @dataclass(config=dataclass_config, repr=False)
 class H3(Heading):
+    """Creates an H3 HTML tag with the text specified.
```
Same as above

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -235,32 +319,54 @@ class Link(Base):
 
 @dataclass(config=dataclass_config, repr=False)
 class InlineLatex(Base):
+    """A block of inline LaTeX. Does not add newline
```
Avoid the term `block` here which is confusing because it has a very specific meaning in Reports.

This is just inline latex

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -235,32 +319,54 @@ class Link(Base):
 
 @dataclass(config=dataclass_config, repr=False)
 class InlineCode(Base):
+    """A block of inline code. Does not add newline
```
same as above

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -235,32 +319,54 @@ class Link(Base):
 
 @dataclass(config=dataclass_config, repr=False)
 class P(Block):
+    """A block that contains a paragraph of text.
```
same as above

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -272,10 +378,17 @@ def from_model(cls, model: internal.ListItem):
 
 @dataclass(config=dataclass_config, repr=False)
 class CheckedListItem(Base):
+    """A list item with a checkbox. Add one ore more `CheckedListItem` within `CheckList`.
```
spelling: `or`

Also the container is called `CheckedList`

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -311,13 +436,15 @@ def to_model(self):
 
 @dataclass(config=dataclass_config, repr=False)
 class List(Block):
+    """A list of items."""
```
clarify as an ordered list

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -427,81 +606,120 @@ def to_model(self):
 
 @dataclass(config=dataclass_config, repr=False)
 class GalleryURL(Base):
+    """A URL to an external resource.
```
clarify as in the gallery

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -639,6 +889,23 @@ class Panel(Base):
 @dataclass(config=dataclass_config, repr=False)
 class PanelGrid(Block):
+    """
+    An HTML block where you can add `Runset` and `Panel` objects to your project.
```
This is not an HTML block

Clarify as a place to add panels and runsets

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -698,23 +965,34 @@ def _validate_list_not_empty(cls, v):  # noqa: N805
-    def to_model(self):
+    """
+    A block that contains a list of sections and subsections using
+    H1, H2, and H3 HTML tags specified in a report.
```
H1-3 blocks, not tags

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -1269,19 +1547,69 @@ def from_model(cls, model: internal.WeaveBlock):
+    A panel object with 2D line plots.
+
+    Attributes:
+        title (Optional[str]): The text that appears at the top of the plot.
```
consistency: Why do you add types here but not above?

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -1379,6 +1707,32 @@ def from_model(cls, model: internal.LinePlot):
+        y Optional[SummaryOrConfigOnlyMetric]:  One or more metrics logged to your W&B project that the report pulls
+            information from. Metrics specified are plotted within the y-axis.
+        z Optional[SummaryOrConfigOnlyMetric]:
+        range_x (tuple): Tuple that specifies the range of the x-axis.
```
tuples usually specify the shape (in this case it's (lb, ub))

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -1563,41 +1962,60 @@ def from_model(cls, model: internal.ScatterPlot):
+    A panel object that compares the code between two different runs.
+
+    Attributes:
+        diff (Required): How to display code differences.
```
consistency: Maybe worth going over whether you want to include:
1. Types
2. "Required"

And then apply that formatting standard to all objs

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -1734,20 +2192,27 @@ def from_model(cls, model: internal.MediaBrowser):
 @dataclass(config=dataclass_config, repr=False)
 class MarkdownPanel(Panel):
+    """
+    A panel that renders a markdown.
```
grammar: renders markdown

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -1756,6 +2221,18 @@ def from_model(cls, model: internal.ScatterPlot):
 @dataclass(config=dataclass_config, repr=False)
 class CustomChart(Panel):
+    """
+    A panel that shows a custom chart. The chart is defined by a query.
```
clarify: defined by a weave expression

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -2287,12 +2772,25 @@ def from_model(cls, model: internal.WeavePanel):
 class Report(Base):
+    """
+    A Data Class that represents a W&B Report. Use the returned object's `blocks` attribute to customize your report.
+    Report Data Class objects do not automatically save. Use the `save()` method to persists changes.
```
Just call it an object, not dataclass

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -2377,7 +2881,8 @@ def url(self):
 
     def save(self, draft: bool = False, clone: bool = False):
-        model = self.to_model()
+        """Persists changes made to Report Data Class object."""
```
Persists changes made to the report

andrewtruong commented on wandb_workspaces/reports/v2/interface.py:
```diff
@@ -2414,15 +2919,28 @@ def save(self, draft: bool = False, clone: bool = False):
     def to_html(self, height: int = 1024, hidden: bool = False) -> str:
-        """Generate HTML containing an iframe displaying this report."""
+        """
+        Generate HTML containing an iframe displaying this report.
```
Worth clarifying this is commonly used in jupyter

andrewtruong commented on wandb_workspaces/workspaces/__init__.py:
````diff
@@ -1,3 +1,12 @@
+"""Python library for programmatically working with Weights & Biases Workspace API.
+
+```python
+# How to import
+import wandb_workspaces.workspaces
+```
````
same as above

andrewtruong commented on wandb_workspaces/workspaces/interface.py:
````diff
@@ -1,3 +1,12 @@
+"""Python library for programmatically working with Weights & Biases Workspace API.
+
+```python
+# How to import
+import wandb_workspaces.workspaces
+```
````
same as above

andrewtruong commented on wandb_workspaces/workspaces/interface.py:
```diff
@@ -28,14 +37,14 @@
 
-def is_internal(k):
+def __is_internal(k):
     return k.startswith("_")
 
 
-def should_show(v):
+def __should_show(v):
```
these should be single, not double underscore

andrewtruong commented on wandb_workspaces/workspaces/interface.py:
```diff
@@ -68,7 +77,13 @@ def _spec(self):
 @dataclass(config=dataclass_config, repr=False)
 class SectionLayoutSettings(Base):
-    """Panel layout settings for a section, typically seen at the top right of the section in the UI."""
+    """Panel layout settings for a section, typically seen at the top right of the section in the UI.
```
Worth clarifying this is the Workspace UI

ngrayluna commented:
@sephmard What do we want to do about these APIs? They refer to the old Weave product and might confuse customers.

 'WeaveBlockArtifact',
 'WeaveBlockArtifactVersionedFile',
 'WeaveBlockSummaryTable',
 'WeavePanel',
 'WeavePanelArtifact',
 'WeavePanelArtifactVersionedFile',
 'WeavePanelSummaryTable',

@andrewtruong Within reports, do we want to expose? Current implementation shows them:

`Block`
`Heading`
`ListItem`
`UnknownBlock`
`UnknownPanel`

ngrayluna committed:
More updates. Incorporate most feedback from AT

ngrayluna committed:
Added weave/query panel info

ngrayluna committed:
Temp fix to remove unwanted classes

ngrayluna commented:
* Decided to leave in `Weave*` and added a note clarifying that it is not releated to LLM Weave. In the future, remove.
* Added "INSERT" for some (what I assume) are Classes that we don't want to show. Re: API generator sees them. In the future, do a) confirm and b) add an underscore so lazydocs doesn't pick them up.

ngrayluna committed:
Incorporated Andrews feedback + added types to workspaces
